Introduced AliasInfo for OpInfo

vfdev-5 opened the pull request:
Introduced AliasInfo for OpInfo.

Context: Split of #49158

cc @mruberry , please let me know if you'd like to see here more code to cover 

> [ ] fold test_op_aliases.py into OpInfo-based testing in test_ops.py

from https://github.com/pytorch/pytorch/issues/50006

and/or add `UnaryUfuncInfo('abs')` as discussed https://github.com/pytorch/pytorch/pull/49158/files#r548774221

Commit:
Introduced AliasInfo for OpInfo
Context: #49158

Commit:
Added arcsin alias, added abs UnaryUfuncInfo

vfdev-5 committed:
Merge branch 'master' into add-opinfo-aliases

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -80,6 +80,19 @@ def _getattr_qual(obj, name, default=_NOTHING):
 class OpInfo(object):
     """Operator information and helper functions for acquiring it."""
 
+    class AliasInfo(object):
```
For consistency with other classes defined in this file, like SampleInput, let's pull this out of OpInfo.

mruberry commented:
Hey @vfdev-5, this is exactly what I was thinking. I think we should extend this PR a little further like you suggest. 

Let's port 

https://github.com/pytorch/pytorch/blob/4a3a37886cdbdca108536a2730c034db4f96f095/test/test_op_aliases.py#L186

and 

https://github.com/pytorch/pytorch/blob/4a3a37886cdbdca108536a2730c034db4f96f095/test/test_op_aliases.py#L248

to test_ops.py and update to be run on OpInfos that have one or more AliasInfos. This is going to be a little tricky because you'll have to separate that test from the custom test generator it's in. The original tests should remain for now, too, until we port all the aliases in test_op_aliases.py to OpInfo metadata. To test that these two new alias tests are working properly we can add an AliasInfo to acos:

https://github.com/pytorch/pytorch/blob/4a3a37886cdbdca108536a2730c034db4f96f095/torch/testing/_internal/common_methods_invocations.py#L742

https://github.com/pytorch/pytorch/blob/4a3a37886cdbdca108536a2730c034db4f96f095/test/test_op_aliases.py#L56

And remove the entries arccos entries from alias_infos. 

Then after we get the new tests working and demonstrate that by updating an existing OpInfo to use the pattern we can follow-up by porting the other alias_infos and delete test_op_aliases.py. Sound good?

vfdev-5 commented:
@mruberry thanks a lot for the review and a thourough description on the improvements ! Sounds great ! Let me update the PR and I'll let you know once it is ready for another review round.

Commit:
[WIP] Added test_aliases_consistency_eager/jit methods to test_ops.TestCommon

vfdev-5 commented on test/test_ops.py:
```diff
@@ -339,6 +339,91 @@ def test_out(self, device, dtype, op):
         self.assertEqual(expected, out)
 
+    @ops([op for op in op_db if op.aliases])
+    def test_aliases_consistency_eager(self, device, dtype, op):
```
Currently, there is a code duplication based on `test_variant_consistency_eager` to check forward only results.

vfdev-5 commented on test/test_ops.py:
```diff
@@ -339,6 +339,91 @@ def test_out(self, device, dtype, op):
+                        self.assertEqual(alias_forward, expected_forward, atol=0, rtol=0)
+
+    @ops([op for op in op_db if op.aliases])
+    def test_aliases_consistency_jit(self, device, dtype, op):
```
Same here about code duplication

Commit:
Merge branch 'master' of https://github.com/pytorch/pytorch into add-opinfo-aliases

Commit:
Fixed test_aliases_consistency_jit

vfdev-5 committed:
Merge branch 'master' into add-opinfo-aliases

Commit:
Fixed failing tests

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -986,7 +1035,8 @@ def sample_inputs_fliplr_flipud(op_info, device, dtype, requires_grad):
                                 dtypes=[torch.cdouble], active_if=IS_WINDOWS),
-                   )),
+                   ),
+                   aliases=['arccos', ]),
```
Nit: "aliases" seems like good metadata to appear early in the parameters initializing the OpInfo. This is fine but maybe put it under the name?

mruberry replied:
Double nit: make it a tuple since it's an immutable iterable

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -168,6 +183,10 @@ def __init__(self,
         self.check_batched_grad = check_batched_grad
         self.check_batched_gradgrad = check_batched_gradgrad
 
+        self.aliases = []
```
Tuples over lists here, too.

mruberry commented on test/test_ops.py:
```diff
@@ -345,6 +345,115 @@ def test_out(self, device, dtype, op):
+        if len(samples) == 0:
+            self.skipTest("Skipped! No sample inputs!")
+
+        for sample in samples:
```
This is cool but this can probably just validate using the first item in samples, instead of all of them.

mruberry commented on test/test_ops.py:
```diff
@@ -345,6 +345,115 @@ def test_out(self, device, dtype, op):
+            expected_forward = op(*sample.input, *sample.args, **sample.kwargs)
+            # Computes expected backward
+            # NOTE: backward may fail for some dtypes
+            exception_during_backwards = False
```
I agree with the comment above that in a future PR we should look at refactoring and creating a helper function for this code.

mruberry commented on test/test_ops.py:
```diff
@@ -345,6 +345,115 @@ def test_out(self, device, dtype, op):
+        if len(samples) == 0:
+            self.skipTest("Skipped! No sample inputs!")
+
+        for sample in samples:
```
First sample only should be OK here, too (computing the forward for all sample inputs for some operators is prohibitively expensive)

mruberry commented on test/test_ops.py:
```diff
@@ -345,6 +345,115 @@ def test_out(self, device, dtype, op):
+                                                    expected_grad, exception_during_backwards)
+
+    @ops([op for op in op_db if op.aliases])
+    def test_aliases_consistency_jit(self, device, dtype, op):
```
This test is good but I don't think it's testing that operator names are converted by the JIT:

https://github.com/pytorch/pytorch/blob/627a33125735a0683083f3e2b56e17fee80741b8/test/test_op_aliases.py#L228

That test is probably sufficient to ensure consistency, since the JIT is actually running the original function once the name is remapped.

What do you think about removing the jit consistency check and replacing it with the jit remapping check, @vfdev-5?

vfdev-5 replied:
Sounds great ! Thanks for pointing out !

mruberry commented on test/test_ops.py:
```diff
@@ -345,6 +345,115 @@ def test_out(self, device, dtype, op):
         self.assertEqual(expected, out)
 
+    @ops([op for op in op_db if op.aliases])
+    def test_aliases_consistency_eager(self, device, dtype, op):
```
What would you think about extending this:

https://github.com/pytorch/pytorch/blob/306420b66e5ab695de99977f481279a9ce764c6e/test/test_ops.py#L203

with the aliases?

Once small challenge with that approach is that it requires updating the inplace checks, like this one:

https://github.com/pytorch/pytorch/blob/627a33125735a0683083f3e2b56e17fee80741b8/test/test_ops.py#L224

to account for inplace aliases, too. This could be done by adding an `is_inplace_variant(self, variant)` method to the OpInfo class that returns True if the variant (including alias variants) is inplace. Or by just checking for `inplace or alias_inplace` (since currently all ops only have one alias). 

This would reduce code duplication and ensure the tests stay consistent in the future.

vfdev-5 replied:
Thanks for the idea ! Yes, such that we shouldn't wait for a refactoring PR. Let me implement that.

mruberry reviewed (commented):
Hey @vfdev-5! This is really cool. All the OpInfo changes look great. I have two questions about the tests which I inline, but can be summarized as:

- for the eager consistency test, should we just add alias variants as variants?
- for the jit consistency test, should we just check that the jit correctly remaps the operation in jit graphs, instead, like the current alias tests do?

Looking forward to hearing your thoughts. This new approach to alias testing (having to just add one alias line in an OpInfo) is way better than the old approach.

Last issue is that this should update this guidance:

https://github.com/pytorch/pytorch/blob/627a33125735a0683083f3e2b56e17fee80741b8/aten/src/ATen/native/native_functions.yaml#L230

to reflect the new approach. Basically I think we should tell people to add the line to an op's OpInfo if the OpInfo exists, and add the OpInfo (and then add the line) if it doesn't exist.

Commit:
Merge branch 'master' of https://github.com/pytorch/pytorch into add-opinfo-aliases

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -960,6 +979,36 @@ def sample_inputs_fliplr_flipud(op_info, device, dtype, requires_grad):
 
 # Operator database (sorted alphabetically)
 op_db: List[OpInfo] = [
+    UnaryUfuncInfo('abs',
```
Since this adds an OpInfo for abs, this PR can remove other generated abs tests:

https://github.com/pytorch/pytorch/blob/869081961824dfb10ee944145c3a8d435571c53b/torch/testing/_internal/common_methods_invocations.py#L2104

https://github.com/pytorch/pytorch/blob/869081961824dfb10ee944145c3a8d435571c53b/test/test_torch.py#L6846

Commit:
Updated code according to PR's review
- jit_name_remapping instead jit consistency
- removed auto generated tests for abs
- added aliases tests inside test_variant_consistency_eager
- added aliases tests to test_out

vfdev-5 commented on test/test_ops.py:
```diff
@@ -339,11 +350,45 @@ def test_out(self, device, dtype, op):
+            _test(a_op)
+
+    @ops([op for op in op_db if op.aliases])
+    def test_aliases_jit_name_remapping(self, device, dtype, op):
```
@mruberry I replaced jit consistency by jit_name_remapping. Please, let me know what do you think. Thanks

mruberry replied:
Sounds great!

Commit:
Fixed typing issue by ignoring the type

mruberry commented on test/test_ops.py:
```diff
@@ -200,7 +200,18 @@ def test_variant_consistency_eager(self, device, dtype, op):
+            aliases = []
+            for a_op in op.aliases:
+                aliases.append(a_op.op)
+                if a_op.method_variant is not None:
```
Because of the filtering that occurs on line 214 I don't think you need to check for None here or on line 209 below. Although that will put None into `inplace_ops` I think that's OK?

vfdev-5 replied:
Yes, you are right about filtering None here and if omitted then `inplace_ops` can be, for example, `[torch_op, None, None]` and yes, it is OK to check `variant` with such list. Just I was thinking that it is not that clean...
I can change as you suggest if you'd like

mruberry commented on test/test_ops.py:
```diff
@@ -339,11 +350,45 @@ def test_out(self, device, dtype, op):
+
+    @ops([op for op in op_db if op.aliases])
+    def test_aliases_jit_name_remapping(self, device, dtype, op):
+        samples = op.sample_inputs(device, dtype, requires_grad=True)
```
This is on the right path. Here's my attempt at annotating a modified version of the current test:

```
#  decorator goes here
def test_jit_alias_remapping(self, device, dtype, op):
  tensor = torch.tensor
  # acquire all op variants and note which are inplace
  for each alias op variant:
    # determine if it's inplace
    # get the string name of the variant (e.g. "absolute", "absolute_")
    # get the corresponding "original name" (e.g. "abs", "abs_")
    # acquire the args from the first sample input

    # Checks that scripting converts aliases
    # NOTE: the code to test scripting must be generated since
    #   scripting does not support splatting args or directly
    #   calling torch.Tensor methods. The following
    #   splats args after the first tensor by inlining them as constants.
    if is_inplace:
      fn_template = '''
        def _fn(t):
          return t.{alias_name}({args})
      '''
      arg_string = ', '.join((str(arg) for arg in info.get_args(device)))  # modify this to use acquired args
      script = fn_template.format(alias_name=info.alias_name, args=arg_string)
    else:
      is_input_tensor_list = isinstance(info.get_input(device), Sequence)
      # For sequence of Tensors, annotate the type to be List[Tensor]
      if is_input_tensor_list:
        fn_template = '''
          def _fn(t: List[Tensor]):
            return op(t{args})
        '''
      else:
        fn_template = '''
          def _fn(t):
           return op(t{args})
          '''
        arg_string = ", " + ', '.join((str(arg) for arg in info.get_args(device)))  # modify this to use args
        script = fn_template.format(args=arg_string)

    # Compiles script
    scripted = torch.jit.CompilationUnit(script)._fn 

    # Acquires and checks the graph remaps the alias
    inp = info.get_input(device)  # this can be acquired from sample inputs now
    # input will have to be cloned for inplace ops. Maybe cloning should be added to SampleInputs since it 
    #  comes up so much? The clone only needs to copy the input tensor.
    scripted(clone_inp(inp))  
    graph = scripted.graph_for(clone_inp(inp))
    FileCheck().check(info.original_name).check_not(info.alias_name).run(graph)
    # Note at this point we're done with SCRIPTING, now we want to check TRACING

    # Checks that tracing converts aliases
    # NOTE: tracing has no problem splatting args
    args = info.get_args(device)  # this will need to be updated to use the args acquired earlier

    def _fn(t, info=info, args=args):
      return info.alias_op(t, *args)

    # the cloned inputs here can be handled like they are above
    traced = torch.jit.trace(_fn, (clone_inp(inp),))
    traced(clone_inp(inp))
    graph = traced.graph_for(clone_inp(inp))
    FileCheck().check(info.original_name).check_not(info.alias_name).run(graph)
```

vfdev-5 replied:
Oh, I see. Thanks! Sorry, I didn't understand that we'd like to port completely the code from test_op_aliases.py.

mruberry commented on test/test_ops.py:
```diff
@@ -339,11 +350,45 @@ def test_out(self, device, dtype, op):
+
+        for a_op in op.aliases:
+
+            def _fn(*sample_args, **sample_kwargs):                
```
This is cool but it's only testing TRACING, not SCRIPTING. See my modified test above, derived from the one in test_op_aliases.py. Basically we want to copy that test and update its data source to use OpInfos instead of its current AliasInfos.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2101,7 +2152,6 @@ def method_tests():
         ('view_as_complex', (S, S, 2), NO_ARGS),
         ('complex', (S, S, S), ((S, S, S),), ''),
-        ('abs', (S, S, S), NO_ARGS, '', (True,)),
         ('abs', (), NO_ARGS, 'scalar', (True,)),
```
This line can go, too.

mruberry replied:
(If we don't have a corresponding unary ufunc sample input let's just add one.)

Commit:
Updates according to the review
- added jit scripting in test_jit_alias_remapping
- etc

mruberry commented on test/test_ops.py:
```diff
@@ -339,11 +349,106 @@ def test_out(self, device, dtype, op):
+            _test(a_op)
+
+    @ops([op for op in op_db if op.aliases])
+    def test_jit_alias_remapping(self, device, dtype, op):
```
@eellison would you take a look at this port of the test from test_op_aliases.py? @vfdev-5 is creating a version of that test that uses OpInfos instead of the custom AliasInfos that test_op_aliases.py uses. The idea is to eventually remove the AliasInfos by adding their metadata to OpInfos.

vfdev-5 committed:
Merge branch 'master' into add-opinfo-aliases

vfdev-5 committed:
Fixed flake8

vfdev-5 commented:
@eellison could you please review this PR in order to unblock other PRs. Thanks !

eellison reviewed (approved):
LGTM

mruberry reviewed (approved):
Cool! Thanks @vfdev-5! It's great to have the architecture to further simplify our test suite by making alias testing as simple as adding a line of metadata to an OpInfo.
